R/update_package_object.R

```diff
@@ -0,0 +1,59 @@
+#' Update a data object and accompanying package resource map
+#'
+#' This function updates an data object, and then automatically
+#' updates the package resource map with the new data pid. It
+#' is a convenience wrapper around \code{\link{arcticdatautils::update_object}}
+#' and \code{\link{arcticdatautils::update_resource_map}}.
+#'
+#' @param mn (MNode) Member node
+#' @param data_pid (character) PID for data object to update
+#' @param new_data_path (character) Path to new data object
+#' @param resource_map_pid (character) PID for resource map to update
+#' @param format_id (character) Optional. The format ID to set for the object.
+#' When not set, \code{\link{arcticdatautils::guess_format_id}} will be used
+#' to guess the format ID. Should be a \href{https://cn.dataone.org/cn/v2/formats}{DataONE format ID}.
+#' @param public (logical) Optional. Make the update public. If FALSE,
+#' will set the metadata and resource map to private (but not the data objects).
+#' This applies to the new metadata PID and its resource map and data object.
+#' Access policies are not affected.
+#'
+#' @keywords update_object update_resource_map
+#'
+#' @export
+
+update_package_object <- function(mn,
+                           data_pid,
+                           new_data_path,
+                           resource_map_pid,
+                           format_id = NULL,
+                           public = FALSE) {
+
+    #argument checks:
+    library(assertthat)
+    library(arcticdatautils)
+    assertthat::assert_that(is(mnTest, "MNode"),
+                is.character(data_pid),
+                is.character(new_data_path),
+                file.exists(new_data_path),
+                is.character(resource_map_pid),
+                is.logical(public))
+
+    new_data_pid <- arcticdatautils::update_object(mn,
+                                  data_pid,
+                                  new_data_path,
+                                  format_id = format_id)
+
+    pkg <- arcticdatautils::get_package(mn, resource_map_pid)
+    other_data_pids <- pkg$data[pkg$data != data_pid]
+    new_data_pids <- c(other_data_pids,
+                       new_data_pid)
+
+    new_resource_map_pid <- arcticdatautils::update_resource_map(mn,
+                        pkg$resource_map,
+                        pkg$metadata,
+                        new_data_pids,
+                        public = public)
+
+    #output package info
+    get_package(mn, new_resource_map_pid)
```

Comment:
Good idea for a return value. Though there is a big potential problem here. `get_package` uses the Solr index which you can't rely on to be correct so soon after a call to `update_resource_map`. Calls to `update_resource_map` ask the Solr index to eventually update the index which will always take longer than the code you execute. We can think about it in a chronological sense:

```
 Time   Action
 1      Your call to update_resource_map
 2      Your call to get_package, which returs out of date info
 3
 4
 5
 6      Solr index updates
 7      When you actually should've called get_package
 8    
 9    
10
```

You can actually reason about what the value should be because you can just return:

```r
list(metadata = pkg$metadata, resource_map = new_resource_map_pid, data = new_data_pids)

```

All that said, maybe there's another option for a return value. What if, say, someone is iteratively editing an EML and wants to keep updating its package. If you return the full package, the user has to figure out what their new PID is each time:

```r
# edit EML, then
new_pkg <- update_package_object(mn, "X", "science_metadata.xml")
# edit EML again, then
new_new_pkg <- update_package_object(mn, "Y", "science_metadata.xml")
# edit EML again, then
new_new_new_pkg <- update_package_object(mn, "Z", "science_metadata.xml")
```

But if you return the new PID, the user can re-run the same line of code over and over again:

```r
# edit EML, then
new_pid <- update_package_object(mn, "X", "science_metadata.xml")
# edit EML again, then
new_pid <- update_package_object(mn, new_pid, "science_metadata.xml")
# edit EML again, then
new_pid <- update_package_object(mn, new_pid, "science_metadata.xml")
```

What do you think? Jeanette was really wanting this to be easy earlier today.

Reply:
Good point!  I didn't consider it at all.  I'll work add in a `get_package` substitute to return the new pids directly